cmd/soroban-cli/src/commands/init.rs

```diff
@@ -0,0 +1,329 @@
+            ExampleContract::Increment => String::from("increment"),
+            ExampleContract::LiquidityPool => String::from("liquidity_pool"),
+            ExampleContract::Logging => String::from("logging"),
+            ExampleContract::SimpleAccount => String::from("simple_account"),
+            ExampleContract::SingleOffer => String::from("single_offer"),
+            ExampleContract::Timelock => String::from("timelock"),
+            ExampleContract::Token => String::from("token"),
+            ExampleContract::UpgradeableContract => String::from("upgradeable_contract"),
+            ExampleContract::None => String::from("none"),
+        }
+    }
+}
+
+#[derive(Parser, Debug, Clone)]
+#[group(skip)]
+pub struct Cmd {
+    pub project_path: String,
+
+    /// optional flag to specify soroban example contracts to include
+    #[arg(short, long, num_args = 1..=20, default_value = "none")]
+    pub with_contract: Vec<ExampleContract>,
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    #[error("Io error: {0}")]
+    CreateDirError(#[from] io::Error),
+
+    // the gix::clone::Error is too large to include in the error enum as is, so we wrap it in a Box
+    #[error("Failed to clone the template repository")]
+    CloneError(#[from] Box<gix::clone::Error>),
+
+    // the gix::clone::fetch::Error is too large to include in the error enum as is, so we wrap it in a Box
+    #[error("Failed to fetch the template repository: {0}")]
+    FetchError(#[from] Box<gix::clone::fetch::Error>),
+
+    #[error("Failed to checkout the template repository: {0}")]
+    CheckoutError(#[from] gix::clone::checkout::main_worktree::Error),
+
+    #[error("Failed to parse Cargo.toml: {0}")]
+    TomlParseError(#[from] TomlError),
+}
+
+const TEMPLATE_URL: &str = "https://github.com/AhaLabs/soroban-init-template.git";
+const SOROBAN_EXAMPLES_URL: &str = "https://github.com/stellar/soroban-examples.git";
+
+impl Cmd {
+    #[allow(clippy::unused_self)]
+    pub fn run(&self) -> Result<(), Error> {
+        println!("Creating a new soroban project at {}", self.project_path);
+        let project_path = Path::new(&self.project_path);
+
+        init(project_path, TEMPLATE_URL, &self.with_contract)?;
+
+        Ok(())
+    }
+}
+
+fn init(
+    project_path: &Path,
+    template_url: &str,
+    with_contracts: &[ExampleContract],
+) -> Result<(), Error> {
+    // create a template temp dir to clone the template repo into
+    let template_dir = tempfile::tempdir()?;
+
+    // clone the template repo into the temp dir
+    clone_repo(template_url, template_dir.path())?;
+
+    // create the project directory and copy the template contents into it
+    std::fs::create_dir_all(project_path)?;
+    copy_contents(template_dir.path(), project_path)?;
+
+    // if there are with-contract flags, include the example contracts
+    if include_example_contracts(with_contracts) {
+        println!("Including example contracts: {with_contracts:?}");
```

Comment:
Can we print a comma separated list of contract names instead of the array?


```suggestion
        println!("Including example contracts: {}", with_contracts.iter().map(ToString::to_string).join(", "));
```

Or, if you replace the `impl ToString for ExampleContract` with a `impl Display for ExampleContract` you'll get the `ToString` impl for free and you'll be able to write the join as:

```suggestion
        println!("Including example contracts: {}", with_contracts.iter().join(", "));
```

_For both you'll need to add `use itertools::Itertools;` to the file._

Reply:
Yep, good call! I'll push a fix up shortly.

Reply:
c10d79a32798b945b0e92f7e308b99e6c739ee31